### DISCUSSION TITLE
422 using a dict body with 0.27.0

### DISCUSSION THREAD

**Discussion by CanD42** [fastapi/fastapi#8122] [Questions] [answered]
The following endpoint was running fine with 0.26.0 and  Python 3.6.5:

@app.patch("/ps/user")
async def patch_user(user_dict: dict, add: List[str] = Query(None)) -> dict:
   ...

but with upgrading to 0.27.0 I got the following 422 error:

```
{
    "detail": [
        {
            "loc": [
                "query",
                "user_dict"
            ],
            "msg": "field required",
            "type": "value_error.missing"
        }
    ]
}
```
I can't find and according change in the last commits so I assume this is a bug?

Thanks,
Daniel

**Comment by tiangolo** [comment, marked as answer]
Thanks! Yep. You caught a bug :bug: 

It should be fixed by #287 . There are also several new tests to check that this behavior doesn't change in the future.

It's available in version `0.27.2` (just released). :tada: :rocket:

**Comment by CanD42** [comment]
wow, that was fast. It's working now with 0.27.2 👍
Thanks

**Comment by tiangolo** [comment]
Thanks for reporting back and closing the issue! :rocket: :cake:

**Comment by DannyVarod** [comment]
@tiangolo, could this issue be back in version 0.61.1?

I'm trying to define an end point like the follow (tried the following and other variations) and all return 422:
(removing the param1 didn't help either)

@app.post('/op1/{param1}')
async def op1(param1: int, my_dict: Dict[Any, Any]):
    pass

@app.post('/op1/{param1}')
async def op1(param1: int, my_dict: Dict[Any, Any] = Body(...)):
    pass

@app.post('/op1/{param1}')
async def op1(param1: int, my_dict: Dict[Any, Any] = Body(default=dict(), embed=False)):
    pass

The only workaround I found was defining my_dict as str and parsing the json myself.

**Comment by Kludex** [comment]
I can't reproduce your issue @DannyVarod 

Run as `pytest <filename>`:
```python
from typing import Dict, Any

import pytest
from fastapi import FastAPI, Body
from fastapi.testclient import TestClient


app = FastAPI()

@app.post('/op1/{param1}')
async def op1(param1: int, my_dict: Dict[Any, Any]):
    return my_dict

@app.post('/op2/{param1}')
async def op2(param1: int, my_dict: Dict[Any, Any] = Body(...)):
   return my_dict

@app.post('/op3/{param1}')
async def op3(param1: int, my_dict: Dict[Any, Any] = Body(default=dict(), embed=False)):
    return my_dict

client = TestClient(app)

@pytest.mark.parametrize("endpoint", ["op1", "op2", "op3"])
def test_everything(endpoint: str):
    param1 = 1
    my_dict = {"potato": "potato"}

    res = client.post(f"/{endpoint}/{param1}", json=my_dict)
    assert res.json() == my_dict
```

**Comment by DannyVarod** [comment]
@Kludex, how comes you used `json=my_dict` and not `data=my_dict` or `json=json.dumps(my_dict)`?

**Comment by tiangolo** [comment]
@DannyVarod the test client is based on Requests, the `json` parameter take pure Python data types and converts them to serialized JSON. The `data` parameter is for form data. `json=json.dumps(my_dict)` would pass a single string to `json`, then Requests would send that as a JSON string type, not as a JSON object.

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by DannyVarod** [comment]
@tiangolo I have since figured this out. The names are misleading though,m because JSON is a form of serializing data into a string - not an object. I would expect data to be for original data and json for **serialized** data (into JSON), incase someone wants to use a custom serializer.

**Comment by tiangolo** [comment]
Yep, it's an old design decision by Requests 🤷😅
